syslog-ng-rs/syslog-ng-common/src/proxies/parser/mod.rs

```diff
@@ -54,58 +56,123 @@ pub mod _parser_plugin {
+                commit_suicide();
+            }
+        }
     }
 
     #[no_mangle]
-    pub extern fn native_parser_proxy_free(_: Box<ParserProxy<$name>>) {
+    pub extern fn native_parser_proxy_free(proxy: Box<ParserProxy<$name>>) {
+        let wrapper = AssertUnwindSafe(proxy);
+
+        match catch_unwind(move || { let _ = wrapper; } ) {
+            Err(error) => {
+                error!("native_parser_proxy_free() panicked, but the panic was caught: {:?}", error);
+                commit_suicide();
+            },
+            Ok(()) => ()
+        }
     }
 
     #[no_mangle]
-    pub extern fn native_parser_proxy_set_option(slf: &mut ParserProxy<$name>, key: *const c_char, value: *const c_char) {
-        let k: String = unsafe { CStr::from_ptr(key).to_owned().to_string_lossy().into_owned() };
-        let v: String = unsafe { CStr::from_ptr(value).to_owned().to_string_lossy().into_owned() };
+    pub extern fn native_parser_proxy_set_option(this: &mut ParserProxy<$name>, key: *const c_char, value: *const c_char) {
+        let mut wrapper_this = AssertUnwindSafe(this);
+        let wrapper_key = AssertUnwindSafe(key);
+        let wrapper_value = AssertUnwindSafe(value);
+
+        let result = catch_unwind(move || {
+            let k: String = unsafe { CStr::from_ptr(*wrapper_key).to_owned().to_string_lossy().into_owned() };
+            let v: String = unsafe { CStr::from_ptr(*wrapper_value).to_owned().to_string_lossy().into_owned() };
+
+            wrapper_this.set_option(k, v);
+        });
 
-        slf.set_option(k, v);
+        match result {
+            Ok(()) => (),
+            Err(error) => {
+                error!("native_parser_proxy_set_option() panicked, but the panic was caught: {:?}", error);
+                commit_suicide();
+            }
+        }
     }
 
     #[no_mangle]
     pub extern fn native_parser_proxy_process(this: &mut ParserProxy<$name>, parent: *mut $crate::sys::LogParser, msg: *mut $crate::sys::LogMessage, input: *const c_char) -> c_int {
-        let input = unsafe { CStr::from_ptr(input).to_str() };
-        let mut parent = LogParser::wrap_raw(parent);
-        let mut msg = LogMessage::wrap_raw(msg);
-        let result = match input {
-            Ok(input) => this.process(&mut parent, &mut msg, input),
-            Err(err) => {
-                error!("{}", err);
-                false
+        let mut wrapper_this = AssertUnwindSafe(this);
+        let wrapper_parent = AssertUnwindSafe(parent);
+        let wrapper_msg = AssertUnwindSafe(msg);
+        let wrapper_input = AssertUnwindSafe(input);
+
+        let result = catch_unwind(move || {
+            let input = unsafe { CStr::from_ptr(*wrapper_input).to_str() };
+            let mut parent = LogParser::wrap_raw(*wrapper_parent);
+            let mut msg = LogMessage::wrap_raw(*wrapper_msg);
+
+            match input {
+                Ok(input) => wrapper_this.process(&mut parent, &mut msg, input),
```

Comment:
I prefer reducing the scope of variable to as small as sensible. For example, I'd move `parent` and `msg` inside this match.